services/shhext/envelopes.go

```diff
@@ -0,0 +1,204 @@
+package shhext
+
+import (
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/status-im/status-go/params"
+	"github.com/status-im/status-go/services/shhext/mailservers"
+	whisper "github.com/status-im/whisper/whisperv6"
+)
+
+type deliveryState struct {
+	state    EnvelopeState
+	expired  int
+	envelope *whisper.Envelope
+}
+
+// EnvelopesMonitor monitors state of the envelopes delivery.
+type EnvelopesMonitor struct {
+	config params.MessageResendConfig
+
+	w                      *whisper.Whisper
+	handler                EnvelopeEventsHandler
+	mailServerConfirmation bool
+
+	mu      sync.Mutex
+	cache   map[common.Hash]*deliveryState
+	batches map[common.Hash]map[common.Hash]struct{}
+
+	mailPeers *mailservers.PeerStore
+
+	wg   sync.WaitGroup
+	quit chan struct{}
+}
+
+// Add hash to a tracker.
+func (mon *EnvelopesMonitor) Add(envelope *whisper.Envelope) {
+	mon.mu.Lock()
+	defer mon.mu.Unlock()
+	mon.cache[envelope.Hash()] = &deliveryState{
+		state:    EnvelopePosted,
+		envelope: envelope,
+	}
+}
+
+func (mon *EnvelopesMonitor) GetState(hash common.Hash) EnvelopeState {
+	mon.mu.Lock()
+	defer mon.mu.Unlock()
+	state, exist := mon.cache[hash]
+	if !exist {
+		return NotRegistered
+	}
+	return state.state
+}
+
+// Start processing events.
+func (mon *EnvelopesMonitor) Start() {
+	mon.quit = make(chan struct{})
+	mon.wg.Add(1)
+	go func() {
+		mon.monitor()
+		mon.wg.Done()
+	}()
+}
+
+// Stop process events.
+func (mon *EnvelopesMonitor) Stop() {
+	close(mon.quit)
+	mon.wg.Wait()
+}
+
+func (mon *EnvelopesMonitor) monitor() {
+	events := make(chan whisper.EnvelopeEvent, 100) // must be buffered to prevent blocking whisper
+	sub := mon.w.SubscribeEnvelopeEvents(events)
+	defer sub.Unsubscribe()
+	for {
+		select {
+		case <-mon.quit:
+			return
+		case event := <-events:
+			mon.process(event)
+		}
+	}
+}
+
+func (mon *EnvelopesMonitor) process(event whisper.EnvelopeEvent) {
+	handlers := map[whisper.EventType]func(whisper.EnvelopeEvent){
+		whisper.EventEnvelopeSent:      mon.handleEventEnvelopeSent,
+		whisper.EventEnvelopeExpired:   mon.handleEventEnvelopeExpired,
+		whisper.EventBatchAcknowledged: mon.handleAcknowledgedBatch,
+	}
+
+	if handler, ok := handlers[event.Event]; ok {
+		handler(event)
+	}
+}
+
+func (mon *EnvelopesMonitor) handleEventEnvelopeSent(event whisper.EnvelopeEvent) {
+	if mon.mailServerConfirmation {
```

Comment:
what if this is `false`?

Reply:
this code was moved due to refactoring. but in general this only ensures that we accept this event only when envelope was sent to mail server